### PR TITLE
Fix some tests and demos to work in both real and complex modes

### PR DESCRIPTION
**Pull request by IgorBaratta**
Key changes:
 - Replace `*` by `inner`
 - Check position of Test and Trial functions in inner products

### CODE REVIEW

**Comment by blechta** [comment]
Should we distinguish `skip_if_complex` and `xfail_if_complex` as a case which is not expected to work and a case which is broken (e.g, IO?)?

File: python/src/common.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ void common(py::module& m)
         "Return `True` if DOLFIN is configured with mpi4py");
   m.def("has_parmetis", &dolfin::has_parmetis);
   m.def("has_scotch", &dolfin::has_scotch);
-  m.def("has_petsc_complex", &dolfin::has_petsc_complex,
+  m.def("complex_mode", &dolfin::has_petsc_complex,
```
Comment (blechta) [inline python/src/common.cpp]: Why is `has_petsc_complex` wrapped as `complex_mode`? This is just confusing. It should be wrapped as `has_petsc_complex`.

File: python/src/common.cpp [hunk 2]
```diff
@@ -57,7 +57,7 @@ void common(py::module& m)
...
         "Return `True` if DOLFIN is configured with PETSc compiled with "
         "scalars represented as complex numbers ");
```
Comment (blechta) [inline python/src/common.cpp]: `Return True if PETSc scalar is complex.`

File: python/test/unit/fem/test_numba_assembly.py [hunk 3]
```diff
@@ -6,17 +6,20 @@
...
 
 from dolfin.jit.jit import ffc_jit
 
 # Define scalar_type here (change to np.complex128 for complex)
-# e.g. if has_petsc_complex(): scalar_type = np.complex128
-scalar_type = np.float64
+if complex_mode():
+    scalar_type = np.complex128
+else:
+    scalar_type = np.float64
```
Comment (blechta) [inline python/test/unit/fem/test_numba_assembly.py]: Is petsc4py required dependency? If yes, use `petsc4py.PETSc.ScalarType`.

File: python/test/unit/fem/test_numba_assembly.py [hunk 4]
```diff
@@ -6,21 +6,18 @@
...
 from numba import cfunc, types, carray, typeof
-import numpy as np
-
 from dolfin.jit.jit import ffc_jit
-
-# Define scalar_type here (change to np.complex128 for complex)
-# e.g. if has_petsc_complex(): scalar_type = np.complex128
-scalar_type = np.float64
+import numpy as np
+from petsc4py.PETSc import ScalarType
```
Comment (blechta) [inline python/test/unit/fem/test_numba_assembly.py]: Can anybody confirm that `petsc4py` is a required dependency? @chrisrichardson @garth-wells

Comment (IgorBaratta) [reply python/test/unit/fem/test_numba_assembly.py]: I assumed that petsc4py is a required dependency due the closed issue [73](https://github.com/FEniCS/dolfinx/issues/73), and it's in the list of required dependencies in [installation.rst](https://github.com/FEniCS/dolfinx/blob/ed9064cb3151f1804e71128f3c46e5cd0f94bd49/python/doc/source/installation.rst).
If it is not, I'll change back to the old code.

Comment (blechta) [reply python/test/unit/fem/test_numba_assembly.py]: I see, I was just confused by a mess which I just reported https://github.com/FEniCS/dolfinx/issues/125

**Pull request merged**